code/obj/mining.dm

```diff
@@ -701,161 +702,82 @@
 		src.updateUsrDialog()
 		return
 
-	proc/generate_interface(var/mob/user as mob)
-		src.add_dialog(user)
-
-		var/dat = "<BR><B>Magnet Status:</B><BR>"
-		dat += "<u>Condition:</u> "
-		switch(src.health)
-			if (95 to INFINITY)
-				dat += "Optimal"
-			if (70 to 94)
-				dat += "Mild Structural Damage"
-			if (40 to 69)
-				dat += "Heavy Structural Damage"
-			if (10 to 39)
-				dat += "Extreme Structural Damage"
-			if (-INFINITY to 10)
-				dat += "Destruction Imminent"
-
-		dat += "<br><u>Status:</u> "
-		if (src.active)
-			dat += "Pulling New Mineral Source"
-		else
-			if (src.last_used > world.time)
-				dat += "Cooling Down: Ready in T-[max(0,round((src.last_used - world.time) / 10))]"
-				if (src.cooldown_override)
-					dat += "<br><i>Cooldown Override Engaged</i>"
-			else
-				dat += "Idle"
+	ui_data(mob/user)
+		. = ..()
+		.["magnetHealth"] = src.health
+		.["magnetActive"] = src.active
+		.["magnetLastUsed"] = src.last_used
+		.["magnetCooldownOverride"] = src.cooldown_override
+		.["magnetAutomaticMode"] = src.automatic_mode
+
+		var/list/miningEncounters = list()
+		for(var/encounter_id in mining_controls.mining_encounters_selectable)
+			var/datum/mining_encounter/encounter = mining_controls.mining_encounters_selectable[encounter_id]
+			if(istype(encounter))
+				miningEncounters += list(list(
+					name = encounter.name,
+					id = encounter_id
+				))
+		.["miningEncounters"] = miningEncounters
+
+		.["time"] = TIME
+
+	ui_act(action, params)
+		var/magnetNotReady = src.active || (src.last_used > TIME && !src.cooldown_override) || src.last_use_attempt > TIME
+		switch(action)
+			if ("geoscan")
+				var/MC = src.get_magnetic_center()
+				if (!MC)
+					boutput(usr, "Error. Magnet is not magnetized.")
+					return
+				mining_scan(MC, usr, src.get_scan_range())
```

Comment:
Sanity check on spamming this button and what that would do. Should we have *some* kind of cooldown (even if it's just 1 second)?

Reply:
The handheld geological scanner doesn't have a cooldown, and this button can't mess up the magnet when spammed like "Activate Magnet/telescope location" can, so not really sure.

Reply:
Handheld one also only affects one tile, I believe? Been a while since I've used it.

Reply:
You can use the handheld one to scan a single tile for more details, but using it in-hand does a scan with the same range as the mining magnet.

Reply:
Fair enough. I view this as a slight problem, but it's not one you're introducing here and can be resolved independently.